src/Tools/ExternalAccess/LiveShare/Classification/RoslynSyntaxClassificationService.cs

```diff
@@ -0,0 +1,192 @@
+            _classificationTypeMap = classificationTypeMap;
+            _experimentationService = experimentationService;
+            _threadingContext = threadingContext;
+        }
+
+        public void AddLexicalClassifications(SourceText text, TextSpan textSpan, ArrayBuilder<ClassifiedSpan> result, CancellationToken cancellationToken)
+        {
+            if (_experimentationService.IsExperimentEnabled(WellKnownExperimentNames.SyntacticExp_Local))
+            {
+                using (new RequestLatencyTracker(SyntacticLspLogger.RequestType.LexicalClassifications))
+                {
+                    _originalService.AddLexicalClassifications(text, textSpan, result, cancellationToken);
+                }
+            }
+            else if (_experimentationService.IsExperimentEnabled(WellKnownExperimentNames.SyntacticExp_Remote))
+            {
+                var documentId = _remoteLanguageServiceWorkspace.GetDocumentIdInCurrentContext(text.Container);
+                var document = _remoteLanguageServiceWorkspace.CurrentSolution.GetDocument(documentId);
+                if (document == null)
+                {
+                    // It is expected that the document cannot be found (e.g. source text comes from preview workspace).
+                    // These cases are already not supported in liveshare, so for now just return.
+                    return;
+                }
+                using (new RequestLatencyTracker(SyntacticLspLogger.RequestType.LexicalClassifications))
+                {
+                    _threadingContext.JoinableTaskFactory.Run(async () =>
+                    {
+                        await AddRemoteClassificationsAsync(LexicalClassificationsHandler.LexicalClassificationsMethodName, document.FilePath, text, textSpan, result, cancellationToken).ConfigureAwait(false);
+                    });
+                }
+            }
+            else
+            {
+                // Some other invalid flight.  Just fallback to the regular service.  Don't want to block the user based on an experimentation failure.
+                _originalService.AddLexicalClassifications(text, textSpan, result, cancellationToken);
+            }
+        }
+
+        public void AddSemanticClassifications(SemanticModel semanticModel, TextSpan textSpan, Workspace workspace, Func<SyntaxNode, ImmutableArray<ISyntaxClassifier>> getNodeClassifiers, Func<SyntaxToken, ImmutableArray<ISyntaxClassifier>> getTokenClassifiers, ArrayBuilder<ClassifiedSpan> result, CancellationToken cancellationToken)
+        {
+            _threadingContext.JoinableTaskFactory.Run(async () =>
+            {
+                var sourceText = await semanticModel.SyntaxTree.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                await AddRemoteClassificationsAsync(RoslynMethods.ClassificationsName, semanticModel.SyntaxTree.FilePath, sourceText, textSpan, result, cancellationToken).ConfigureAwait(false);
```

Comment:
i'm confused.  why are we getting the text locally, to ask the remote host what the classifications are?  Why can't it just compute this with the text it has?  

What happens here if this is a very large file?